File: ruby/lib/dec_to_base.rb
# frozen_string_literal: true

# https://github.com/joncalhoun/algorithmswithgo.com/tree/master/module01#09---decimal-to-another-base-code

# DecToBase :
class DecToBase
  CHARSET = '0123456789ABCDEF'

  def self.iterate(dec, base)
    res = ''
    loop do
      div = dec / base
      break if dec.zero?

      res = CHARSET[dec % base] + res
      dec = div
    end
    res
  end

  def self.recursive(dec, base)
    div = dec / base
    res = recursive(div, base) if div != 0
    res.to_s + CHARSET[dec % base]
  end

  def self.std(dec, base)
    dec.to_s(base).upcase
  end
end

File: rust/src/dec_to_base.rs
use radix_fmt::radix;

// https://github.com/joncalhoun/algorithmswithgo.com/tree/master/module01#09---decimal-to-another-base-code

const CHARSET: &str = "0123456789ABCDEF";

pub fn iterate(mut dec: u32, base: u32) -> String {
	let mut res = "".to_string();

	loop {
		let div = dec / base;
		if dec == 0 {
			break;
		}
		res = format!(
			"{}{}",
			CHARSET.chars().nth((dec % base) as usize).unwrap(),
			res
		);
		dec = div;
	}
	res
}

pub fn recursive(dec: u32, base: u32) -> String {
	let mut res = "".to_string();
	let div = dec / base;
	if div != 0 {
		res = recursive(div, base)
	}
	format!(
		"{}{}",
		res,
		CHARSET.chars().nth((dec % base) as usize).unwrap()
	)
}

pub fn std(num: u32, base: u32) -> String {
	format!("{}", radix(num, base as u8)).to_uppercase()
}
